### PR TITLE
WIP: Dimension Reduction Update

### PR DESCRIPTION
**Pull request by bkmgit**
Draft for dimension reduction

### CODE REVIEW

File: _episodes/05-dimensionality-reduction.md [hunk 1]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+from sklearn import decomposition
+from sklearn import datasets
+from sklearn import manifold
+
+digits = datasets.load_digits()
+
+# Examine the dataset
+
+print(digits.data)
+
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:43]: Remove extraneous space

File: _episodes/05-dimensionality-reduction.md [hunk 2]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+print(digits.target)
+
+X = digits.data
+y = digits.target
+
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:48]: Remove extraneous space

File: _episodes/05-dimensionality-reduction.md [hunk 3]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+~~~
+{: .python}
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: This should be replaced with `{: .language-python}`

File: _episodes/05-dimensionality-reduction.md [hunk 4]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+X_pca = pca.transform(X)
+
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_pca[:, 0], X_pca[:, 1], c=y, cmap=plt.cm.nipy_spectral, 
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("pca.svg")
+~~~
+{: .python}
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace with `{: .language-python}`

File: _episodes/05-dimensionality-reduction.md [hunk 5]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+        random_state = 0)
+X_tsne = tsne.fit_transform(X)
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_tsne[:, 0], X_tsne[:, 1], c=y, cmap=plt.cm.nipy_spectral,
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("tsne.svg")
+~~~
+{: .python}
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace with `{: .language-python}`

File: _episodes/05-dimensionality-reduction.md [hunk 6]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+> > pca.fit(X)
+> > X_pca = pca.transform(X)
+> > fig = plt.figure(1, figsize=(4, 4))
+> > plt.clf()
+> > ax = fig.add_subplot(projection='3d')
+> > ax.scatter(X_pca[:, 0], X_pca[:, 1], X_pca[:, 2], c=y,
+> >           cmap=plt.cm.nipy_spectral, s=9, lw=0)
+> > plt.savefig("pca_3d.svg")
+> > ~~~
+> > {. :python}
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace with `{: .language-python}`

File: _episodes/05-dimensionality-reduction.md [hunk 7]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+> >         random_state = 0)
+> > X_tsne = tsne.fit_transform(X)
+> > fig = plt.figure(1, figsize=(4, 4))
+> > plt.clf()
+> > ax = fig.add_subplot(projection='3d')
+> > ax.scatter(X_tsne[:, 0], X_tsne[:, 1], X_tsne[:, 2], c=y,
+> >           cmap=plt.cm.nipy_spectral, s=9, lw=0)
+> > plt.savefig("tsne_3d.svg")
+> > ~~~
+> > {. :python}
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace with `{: .language-python}`

File: _episodes/05-dimensionality-reduction.md [hunk 8]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+> >
+> > ![Reduction to 3 components using tsne](../fig/tsne_3d.svg)
+> >
+> >
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:137]: Add `{: .solution}` tag at the end of the double `>` block.

File: _episodes/05-dimensionality-reduction.md [hunk 9]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+>
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Add `{: .challenge}` tag the end of the single `>` block.

**Review by vinisalazar — CHANGES_REQUESTED** [review]
Thank you for your contribution @bkmgit! Your effort is much appreciated.

I requested some minor changes on the tags that you use for the code blocks. If they aren't correctly formatted, the syntax highlighting doesn't work. It would also be nice to have Challenge and Solution tags for the exercises you proposed. 

Let me know if we can help you with anything.

File: _episodes/05-dimensionality-reduction.md [hunk 10]
```diff
@@ -8,10 +8,145 @@ objectives:
...
+import matplotlib.pyplot as plt
+
+from sklearn import decomposition
+from sklearn import datasets
+from sklearn import manifold
+
+digits = datasets.load_digits()
+
+# Examine the dataset
+
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:41]: Remove extraneous space

File: _episodes/05-dimensionality-reduction.md (hunk 1 above)
Comment (bkmgit) [reply _episodes/05-dimensionality-reduction.md:43]: Thanks.

File: _episodes/05-dimensionality-reduction.md (hunk 2 above)
Comment (bkmgit) [reply _episodes/05-dimensionality-reduction.md:48]: Thanks.

**Comment by bkmgit** [comment]
@vinisalazar Let me know if ok

**Review by vinisalazar — APPROVED** [review]
Thank you for contribution. I understand there are the following additional changes:
- Fix tags in other episodes and setup page.
- Add `Gemfile`.
- Rename files that were incorrectly numbered.

These are welcome changes, and although it'd be preferable to have them in separate PRs, given that this lesson is in an early development stage, I have no problem in including them in this one.

@colinsauze LGTM, what do you think?

**Comment by bkmgit** [comment]
Ok. Will start new branch.

File: _episodes/05-dimensionality-reduction.md [hunk 11]
```diff
@@ -3,15 +3,156 @@ title: "Dimensionality Reduction"
...
+> > ~~~
+> > {: .language-python}
+> >
+> > ![Reduction to 3 components using tsne](../fig/tsne_3d.svg)
+> >
+> >
+> {: .solution}
+{: .challenge}
+
+> # Exercise: Parameters
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:141]: I took a closer look and this line must be replaced with `> ## Exercise: Parameters`, that is, the header must have two `##` and not one `#`.

File: _episodes/05-dimensionality-reduction.md [hunk 12]
```diff
@@ -3,15 +3,156 @@ title: "Dimensionality Reduction"
...
+>
+> Look up parameters that can be changed in PCA and t-SNE,
+> and experiment with these. How do they change your resulting
+> plots?  Might the choice of parameters lead you to make different
+> conclusions about your data?
+{: .challenge}
+
+> # Exercise: Other Algorithms
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md:149]: Replace `> # Exercise:` for `> ## Exercise:`

File: _episodes/05-dimensionality-reduction.md [hunk 13]
```diff
@@ -3,15 +3,156 @@ title: "Dimensionality Reduction"
...
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("tsne.svg")
+~~~
+{: .language-python}
+
+![Reduction using t-SNE](../fig/tsne.svg)
+
+
+
+> # Exercise: Working in three dimensions
```
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: I took a closer look and this line must be replaced with `> ## Exercise: Parameters`, that is, the header must have two `##` and not one `#`.

File: _episodes/05-dimensionality-reduction.md (hunk 11 above)
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace `> # Exercise:` for `> ## Exercise:`

File: _episodes/05-dimensionality-reduction.md (hunk 12 above)
Comment (vinisalazar) [inline _episodes/05-dimensionality-reduction.md]: Replace `> # Exercise:` for `> ## Exercise:`

**Review by vinisalazar — CHANGES_REQUESTED** [review]
Sorry, I took a closer look and there's one last change: using double `##` instead of single `#` in Exercise headers, or they won't render correctly.

File: _episodes/05-dimensionality-reduction.md (hunk 13 above)
Comment (bkmgit) [reply _episodes/05-dimensionality-reduction.md]: ok

File: _episodes/05-dimensionality-reduction.md (hunk 12 above)
Comment (bkmgit) [reply _episodes/05-dimensionality-reduction.md]: ok

File: _episodes/05-dimensionality-reduction.md (hunk 11 above)
Comment (bkmgit) [reply _episodes/05-dimensionality-reduction.md]: ok

**Comment by bkmgit** [comment]
At the moment, local build fails as some of the jekyll files have not been updated in a while

**Review by vinisalazar — APPROVED** [review]
Thanks.

**Comment by vinisalazar** [comment]
As for the Jekyll files: we are using the remote theme, so the Jekyll build should mostly be ok because of that. Could you provide more details of what is failing for you?

**Comment by bkmgit** [comment]
Ok, build of current site works. My fork still has some old code.

**Comment by bkmgit** [comment]
Work is now in #16

**Pull request closed without merging**